crates/wasmtime/src/engine.rs

```diff
@@ -783,6 +783,34 @@ impl Engine {
         )
     }
 
+    /// Loads a `CodeMemory` from the specified memory region without copying
+    ///
+    /// The `expected` marker here is whether the bytes are expected to be
+    /// a precompiled module or a component.  The memory pointed to by `size`
+    /// bytes following `ptr` is expected to be a serialized module (.cwasm)
+    /// generated by `[Module::serialize]` or [`Engine::precompile_module] or
+    /// their `Component` counterparts [`Component::serialize`] or
+    /// `[Engine::precompile_component]`.
+    ///
+    /// # Safety
+    ///
+    /// The memory provided must be valid for `size` bytes following `ptr`.
+    /// As there is no copy here, the runtime will be making direct use
+    /// of the provided memory and may read or write to this memory.  As such,
+    /// any outside writes to this memory region will result in undefined and
+    /// likely very undesirable behavior.
+    pub(crate) unsafe fn load_code_raw(
+        &self,
+        ptr: *const u8,
+        size: usize,
```

Comment:
A small bikeshed on this, one thing we could do here is take `NonNull<[u8]>` which (a) bundles ptr/len into one, and (b) helps clarify the contract that it must be non-null

Reply:
I'm definitely not against this; I think the counterpoint is that, at least with the type signature, you lose the ability to document that the usage of the memory is going to be immutable (which I do believe would be an important invariant for the intended use case) and we end up pushing the `*mut` transmute to callers who have a reference to an immutable chunk of memory.

So, we gain the ability to model with types that we require a non-null pointer but we lose the ability to inform the caller that we're using the memory immutably. 🤷

Reply:
That's already true as soon as you're using raw pointers instead of safe references though, so we are already in that world here.

Reply:
Pushed a new commit making the change to `NonNull<[u8]>`.